### PR TITLE
feat(duckdb): add support for read_mysql

### PR DESCRIPTION
**Pull request by ncclementi**
- Closes https://github.com/ibis-project/ibis/issues/8109

### CODE REVIEW

File: ibis/backends/duckdb/tests/test_register.py [hunk 1]
```diff
@@ -184,6 +184,27 @@ def test_read_postgres(con, pgurl):  # pragma: no cover
...
+    s = ibis.schema(dict(x="float64", y="str"))
+
+    mysqlcon.create_table("duckdb_test", df, schema=s, overwrite=True)
+    yield mysqlcon.con
+    mysqlcon.drop_table("duckdb_test", force=True)
+
+
+def test_read_mysql(con, mysqlurl):  # pragma: no cover
+    table = con.read_mysql(
+        f"mysql://{mysqlurl.user.decode()}:{mysqlurl.password.decode()}@{mysqlurl.host}:{mysqlurl.port}/ibis_testing",
```
Comment (ncclementi) [inline ibis/backends/duckdb/tests/test_register.py:201]: @cpcloud  here` mysqlurl.host = localhost` and it looks liek when this goes into the `query_con` that gets run with `_safe_raw_sql` in `read_mysql` in my mac it complains, while if I provide `127.0.0.1` I have no problems.

**Comment by ncclementi** [comment]
Well looks like we triggered the `"localhost"` issue in CI too see failure [here](https://github.com/ibis-project/ibis/actions/runs/8286659965/job/22677178632?pr=8656#step:15:297)

```python-traceback
E               pymysql.err.OperationalError: (2003, "Can't connect to MySQL server on 'localhost' ([Errno 111] Connection refused)")

../../../.cache/pypoetry/virtualenvs/ibis-framework-2c2VoWo5-py3.9/lib/python3.9/site-packages/pymysql/connections.py:711: OperationalError
```

**Comment by cpcloud** [comment]
I opened duckdb/duckdb_mysql#44 to track the localhost issue.

File: ibis/backends/duckdb/__init__.py [hunk 2]
```diff
@@ -988,6 +988,47 @@ def read_postgres(
 
         return self.table(table_name)
 
+    ########
+    def read_mysql(
+        self,
+        uri: str,
+        catalog: str,
```
Comment (ncclementi) [inline ibis/backends/duckdb/__init__.py:995]: @gforsyth I don't know if this is the best choice of word. Talking to Phillip he mentioned that these aliases are registered as catalogs, hence the choice, but feel free to suggest something different.

**Comment by ncclementi** [comment]
I understand why the DuckDB job fails, we are not bring the container up, we can do something similar to what we are doing in `read_postgres` by setting up an env variable and not run the test in CI. 

But I'm confused about why the nix one is failing, help appreciated.

File: ibis/backends/duckdb/__init__.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/duckdb/__init__.py:994]: `catalog` seems like right thing here given the direction we're going with naming those sorts of things.

File: ibis/backends/duckdb/__init__.py [hunk 3]
```diff
@@ -988,6 +988,46 @@ def read_postgres(
 
         return self.table(table_name)
 
+    def read_mysql(
+        self,
+        uri: str,
```
Comment (cpcloud) [inline ibis/backends/duckdb/__init__.py:993]: Can we make all of these arguments required-keyword?

File: ibis/backends/duckdb/tests/test_register.py [hunk 4]
```diff
@@ -177,13 +177,48 @@ def pgurl():  # pragma: no cover
...
+    s = ibis.schema(dict(x="float64", y="str"))
+
+    mysqlcon.create_table("duckdb_test", df, schema=s, overwrite=True)
+    yield mysqlcon.con
+    mysqlcon.drop_table("duckdb_test", force=True)
+
+
+@pytest.mark.skipif(
+    os.environ.get("DUCKDB_MYSQL") is None, reason="avoiding CI shenanigans"
+)
```
Comment (cpcloud) [inline ibis/backends/duckdb/tests/test_register.py:204]: I'll put up a PR to do this with markers. That way we can avoid mystery environment variables.

**Review by cpcloud — APPROVED** [review]
LGTM!

File: ibis/backends/duckdb/tests/test_register.py (hunk 4 above)
Comment (cpcloud) [reply ibis/backends/duckdb/tests/test_register.py:204]: Hm nevermind, this doesn't seem worth the effort right now.

File: ibis/backends/duckdb/__init__.py (hunk 3 above)
Comment (ncclementi) [reply ibis/backends/duckdb/__init__.py:993]: Do you mean not, by not allowing table to be None, and trigger the Value Error later? 

Should I remove that from the postgres and sqlite cases too?

**Comment by ncclementi** [comment]
So, this was fixed https://github.com/duckdb/duckdb_mysql/issues/44, but we can't use it unless we install the extension from the nightlies, so we'll have to wait a bit.

**Pull request merged**